Add machine readable inlining reports

Gbury opened the pull request:
This PR adds the ability for ocamlopt to output marshalled inlining reports, using the `-inlining-report-mr` option. Such marshalled reports can then be read using the small library in https://github.com/Gbury/ocir . Additionally, the linked repo contains an executable that reads a marshalled inlining report and then print it in a human readable format. In the long term, `ocir` will also contain facilities to compare inlining reports.

This patch should be mergeable, but the name of the option could be changed for something a little bit more understandable/intuitive if there's any good proposition; the extension of marshalled files (currently `.obj.magic`) could also be changed if something more standard exists.

mshinwell commented on middle_end/flambda/inlining/inlining_report.ml:
```diff
@@ -136,12 +142,22 @@ let record_decision ~dbg decision =
-  end
+  end;
+  if !Clflags.inlining_report_serialized then begin
+    let ch = open_out_bin (output_prefix ^ ".inlining.obj.magic") in
```
This filename doesn't match the help text for the command-line option (ditto for the human-readable inlining reports, above).  Also, maybe we could just use "`.inlining" for the machine-readable ones?

Gbury replied:
I changed the extension to `.inlining` for the marshalled reports and adjusted the option descriptions to match the actual filenames generated.

mshinwell commented on middle_end/flambda/inlining/inlining_report.ml:
```diff
@@ -136,12 +142,22 @@ let record_decision ~dbg decision =
+    let report : report = `Flambda2_1_0_0 (metadata, Lazy.force l) in
+    Marshal.to_channel ch report [];
+    close_out ch
+  end;
```
I would put a try-with around this whole section; if any exception is raised, we can then print "Warning: inlining report writing failed: <exn>" or similar.

Gbury replied:
Done (using `Misc.try_finally`). Is there a better way to print the warning than `eprintf`, considering that this is not a numbered warning and thus the `Warnings` module does not seem to help much ?

mshinwell replied:
Not that I know of, `eprintf` will be fine for now.

mshinwell commented on driver/main_args.ml:
```diff
@@ -167,9 +167,16 @@ let mk_inline_toplevel f =
+      round) showing the inliner's decisions, in a human readable format"
 ;;
 
+let mk_inlining_report_serialized f =
```
Let's not use "serialized" as the spelling differs between UK/US English.  What about "`-inlining-report-bin`"?

Gbury replied:
changed to `-inlining-report-bin`

mshinwell reviewed (changes requested):
Just a few minor things to fix.

Gbury commented:
@mshinwell  I addressed your reviews

With the last commit, everything is in order and once the PR on the flambda-backend repo is merged (cf https://github.com/ocaml-flambda/flambda-backend/pull/2 ), comparison of inlining reports between flambda1 and flambda2 will be possible.

For instance, consider the following file:
```ocaml
let rec iter f = function
  | [] -> ()
  | x :: r -> f x; iter f r

let main l =
  (iter[@specialised][@unrolled 1]) print_int l
```

Compiling with flambda1 and flambda2 using the `-inlining-report-bin` option to get the inlining reports, and then running ocir on the two files (with a simple `ocir flambda1-report flambda2-report`), yields the following comparison:

```
** Definition of main/23:main_1{inline.ml:6,9--60}
*** Specialization of iter/8->iter/45::∅{inline.ml:7,2--47}
    
    Specialized on the left but not on the right

*** Application of iter/45::iter_0{inline.ml:7,2--47}
    
    Inlined on the left but not on the right

Summary:
  * left only inlinings : 1
  * left only specializations : 1
  * right only inlinings : 0
  * right only specializations : 0
```
